File: Reference/tests/Brontide.Reference.Studio.Tests/ComponentPolicyDistributionWireTests.cs
using System.Net;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text.Json;

namespace Brontide.Reference.Studio.Tests;

public sealed partial class ComponentBindingIntegrationTests
{
    private sealed class Cbi40Handler(
        Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> send) : HttpMessageHandler
    {
        public int Attempts { get; private set; }
        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            Attempts++;
            return send(request, cancellationToken);
        }
    }

    private sealed class Cbi40UnknownLengthContent(byte[] bytes) : HttpContent
    {
        protected override Task SerializeToStreamAsync(Stream stream, TransportContext? context) =>
            stream.WriteAsync(bytes).AsTask();
        protected override bool TryComputeLength(out long length) { length = 0; return false; }
        protected override Task<Stream> CreateContentReadStreamAsync() =>
            Task.FromResult<Stream>(new MemoryStream(bytes, false));
    }

    private sealed record Cbi40Observation(string Code, bool Update, int Attempts, string? Sha256 = null);

    private static ProviderPublisherTrustPolicyDistributionRequest Cbi40Request() =>
        new(new string('A', 64), 0, null);

    private static ProviderPublisherTrustPolicyDistributionResponse Cbi40Response(
        ProviderPublisherTrustPolicyUpdate? update = null) =>
        new(new string('A', 64), 0, null, 1_800_000_000, 1_800_000_060, update,
            "ECDSA-P256-SHA256", "endpoint-key", "endpoint-signature");

    private static ProviderPublisherTrustPolicyUpdate Cbi40Update()
    {
        ProviderPublisherTrustEntry[] entries = [new(
            ProviderPublisherKeyId.Create(new string('A', 64)), ProviderPublisherTrustDisposition.Admitted)];
        return new(1, null, new(ProviderPublisherTrustPolicyIdentity.Compute(entries), entries),
            "ECDSA-P256-SHA256", "authority-key", "authority-signature");
    }

    private static async Task<Cbi40Observation> Cbi40RunAsync(JsonElement vector)
    {
        var mutation = vector.GetProperty("mutation").GetString()!;
        var update = Cbi40Update();
        if (mutation == "request")
        {
            var bytes = ProviderPublisherTrustPolicyDistributionWireCodec.EncodeRequest(Cbi40Request());
            var decoded = ProviderPublisherTrustPolicyDistributionWireCodec.DecodeRequest(bytes);
            return new(decoded == Cbi40Request() ? "request-roundtrip" : "wire-invalid", false, 0,
                Convert.ToHexString(SHA256.HashData(bytes)));
        }
        if (mutation is "current" or "update")
        {
            var bytes = ProviderPublisherTrustPolicyDistributionWireCodec.EncodeResponse(
                Cbi40Response(mutation == "update" ? update : null));
            var decoded = ProviderPublisherTrustPolicyDistributionWireCodec.DecodeResponse(bytes);
            return new(decoded.Update is null ? "response-current" : "response-update", decoded.Update is not null, 0,
                Convert.ToHexString(SHA256.HashData(bytes)));
        }
        if (mutation is "truncated" or "trailing" or "marker" or "utf8")
        {
            var bytes = ProviderPublisherTrustPolicyDistributionWireCodec.EncodeResponse(Cbi40Response());
            bytes = mutation switch
            {
                "truncated" => bytes[..(bytes.Length / 2)],
                "trailing" => [.. bytes, 0],
                _ => bytes,
            };
            if (mutation == "marker") bytes[4] ^= 1;
            if (mutation == "utf8") bytes[4] = 0xFF;
            try
            {
                ProviderPublisherTrustPolicyDistributionWireCodec.DecodeResponse(bytes);
                return new("wire-accepted", false, 0);
            }
            catch (InvalidDataException) { return new("wire-invalid", false, 0); }
        }

        var endpoint = new Uri("https://policy.example.test/v1/trust");
        var encoded = ProviderPublisherTrustPolicyDistributionWireCodec.EncodeResponse(Cbi40Response());
        var handler = new Cbi40Handler(async (request, cancellationToken) =>
        {
            if (mutation == "canceled") await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken);
            var requestBytes = await request.Content!.ReadAsByteArrayAsync(cancellationToken);
            Assert.Multiple(() =>
            {
                Assert.That(request.Method, Is.EqualTo(HttpMethod.Post));
                Assert.That(request.RequestUri, Is.EqualTo(endpoint));
                Assert.That(request.Content.Headers.ContentType?.MediaType,
                    Is.EqualTo(HttpProviderPublisherTrustPolicyDistributionSource.MediaType));
                Assert.That(request.Headers.Accept.Single().MediaType,
                    Is.EqualTo(HttpProviderPublisherTrustPolicyDistributionSource.MediaType));
                Assert.That(ProviderPublisherTrustPolicyDistributionWireCodec.DecodeRequest(requestBytes),
                    Is.EqualTo(Cbi40Request()));
            });
            HttpContent content = mutation switch
            {
                "declared-oversize" => new ByteArrayContent(new byte[
                    ProviderPublisherTrustPolicyDistributionWireCodec.MaximumMessageBytes + 1]),
                "streamed-oversize" => new Cbi40UnknownLengthContent(new byte[
                    ProviderPublisherTrustPolicyDistributionWireCodec.MaximumMessageBytes + 1]),
                _ => new ByteArrayContent(encoded),
            };
            content.Headers.ContentType = new MediaTypeHeaderValue(
                mutation == "content-type" ? "application/octet-stream"
                    : HttpProviderPublisherTrustPolicyDistributionSource.MediaType);
            if (mutation == "content-encoding") content.Headers.ContentEncoding.Add("gzip");
            return new HttpResponseMessage(mutation == "status" ? HttpStatusCode.ServiceUnavailable : HttpStatusCode.OK)
            {
                RequestMessage = new HttpRequestMessage(HttpMethod.Post,
                    mutation == "redirect" ? new Uri("https://other.example.test/v1/trust") : endpoint),
                Content = content,
            };
        });
        using var client = new HttpClient(handler);
        var source = new HttpProviderPublisherTrustPolicyDistributionSource(client, endpoint);
        using var cancellation = new CancellationTokenSource();
        if (mutation == "canceled") cancellation.Cancel();
        try
        {
            var response = await source.FetchAsync(Cbi40Request(), cancellation.Token);
            return new("transport-success", response.Update is not null, handler.Attempts);
        }
        catch (OperationCanceledException) { return new("transport-canceled", false, handler.Attempts); }
        catch (InvalidDataException) { return new("transport-invalid", false, handler.Attempts); }
    }

    private static JsonDocument Cbi40Fixture() => JsonDocument.Parse(File.ReadAllText(Path.Combine(
        TestContext.CurrentContext.TestDirectory, "component-management", "fixtures",
        "cbi40-policy-distribution-wire-vectors.json")));

    [Test]
    public async Task Shared_cbi40_vectors_encode_and_transport_only_strict_bounded_messages()
    {
        using var fixture = Cbi40Fixture();
        foreach (var vector in fixture.RootElement.GetProperty("vectors").EnumerateArray())
        {
            var actual = await Cbi40RunAsync(vector);
            Assert.Multiple(() =>
            {
                Assert.That(actual.Code, Is.EqualTo(vector.GetProperty("code").GetString()));
                Assert.That(actual.Update, Is.EqualTo(vector.GetProperty("update").GetBoolean()));
                Assert.That(actual.Attempts, Is.LessThanOrEqualTo(1));
                if (vector.TryGetProperty("sha256", out var digest))
                    Assert.That(actual.Sha256, Is.EqualTo(digest.GetString()));
            });
        }
    }

    [Test]
    public void Cbi40_C1_request_codec_is_canonical_complete_and_strict()
    {
        var bytes = ProviderPublisherTrustPolicyDistributionWireCodec.EncodeRequest(Cbi40Request());
        Assert.That(ProviderPublisherTrustPolicyDistributionWireCodec.EncodeRequest(
            ProviderPublisherTrustPolicyDistributionWireCodec.DecodeRequest(bytes)), Is.EqualTo(bytes));
        Assert.Throws<InvalidDataException>(() =>
            ProviderPublisherTrustPolicyDistributionWireCodec.DecodeRequest([.. bytes, 0]));
    }

    [Test]
    public void Cbi40_C2_response_codec_preserves_the_complete_optional_update()
    {
        using var authority = ECDsa.Create(ECCurve.NamedCurves.nistP256);
        var update = Cbi37Sign(authority, 1, null, Cbi37Policy(false));
        var decoded = ProviderPublisherTrustPolicyDistributionWireCodec.DecodeResponse(
            ProviderPublisherTrustPolicyDistributionWireCodec.EncodeResponse(Cbi40Response(update)));
        Assert.That(ProviderPublisherTrustPolicyDistributionManifest.UpdateDigest(decoded.Update!),
            Is.EqualTo(ProviderPublisherTrustPolicyDistributionManifest.UpdateDigest(update)));
    }

    [Test]
    public async Task Cbi40_C3_malformed_truncated_and_trailing_messages_are_refused()
    {
        using var fixture = Cbi40Fixture();
        foreach (var index in new[] { 3, 4, 5, 6 })
            Assert.That((await Cbi40RunAsync(fixture.RootElement.GetProperty("vectors")[index])).Code,
                Is.EqualTo("wire-invalid"));
    }

    [Test]
    public void Cbi40_C4_transport_requires_an_exact_https_endpoint()
    {
        using var client = new HttpClient(new Cbi40Handler((_, _) => throw new AssertionException("not called")));
        Assert.Throws<ArgumentException>(() =>
            new HttpProviderPublisherTrustPolicyDistributionSource(client, new Uri("http://policy.example.test")));
    }

    [Test]
    public async Task Cbi40_C5_http_metadata_and_streaming_are_bounded_without_retry()
    {
        using var fixture = Cbi40Fixture();
        var vectors = fixture.RootElement.GetProperty("vectors");
        foreach (var index in new[] { 7, 8, 9, 10, 11, 13, 14 })
            Assert.That((await Cbi40RunAsync(vectors[index])).Attempts, Is.EqualTo(1));
    }

    [Test]
    public async Task Cbi40_C6_https_source_composes_with_authenticated_durable_distribution()
    {
        var root = Path.Combine(Path.GetTempPath(), $"brontide-cbi40-{Guid.NewGuid():N}");
        try
        {
            using var authority = ECDsa.Create(ECCurve.NamedCurves.nistP256);
            using var endpointKey = ECDsa.Create(ECCurve.NamedCurves.nistP256);
            var authorityId = ProviderPublisherTrustPolicyAuthorityId.Create(
                Convert.ToHexString(SHA256.HashData(authority.ExportSubjectPublicKeyInfo())));
            var endpointId = ProviderPublisherTrustPolicyDistributionEndpointId.Create(
                Convert.ToHexString(SHA256.HashData(endpointKey.ExportSubjectPublicKeyInfo())));
            var durable = DurableProviderPublisherTrustPolicyRegistry.Open(
                Path.Combine(root, "policy.checkpoint"), authorityId).Registry!;
            var endpoint = new Uri("https://policy.example.test/v1/trust");
            var now = DateTimeOffset.FromUnixTimeSeconds(1_800_000_000);
            var update = Cbi37Sign(authority, 1, null, Cbi37Policy(false));
            var handler = new Cbi40Handler(async (request, cancellationToken) =>
            {
                var decoded = ProviderPublisherTrustPolicyDistributionWireCodec.DecodeRequest(
                    await request.Content!.ReadAsByteArrayAsync(cancellationToken));
                var response = Cbi39Respond("update", decoded, endpointKey, update, now);
                return new HttpResponseMessage(HttpStatusCode.OK)
                {
                    RequestMessage = new HttpRequestMessage(HttpMethod.Post, endpoint),
                    Content = new ByteArrayContent(ProviderPublisherTrustPolicyDistributionWireCodec.EncodeResponse(response))
                    { Headers = { ContentType = new MediaTypeHeaderValue(HttpProviderPublisherTrustPolicyDistributionSource.MediaType) } },
                };
            });
            using var client = new HttpClient(handler);
            var source = new HttpProviderPublisherTrustPolicyDistributionSource(client, endpoint);
            var result = await new ProviderPublisherTrustPolicyDistributionClient(durable, endpointId)
                .SynchronizeAsync(source, now, TimeSpan.FromSeconds(1));
            Assert.Multiple(() =>
            {
                Assert.That(result.IsApplied, Is.True);
                Assert.That(result.Floor.Sequence, Is.EqualTo(1));
                Assert.That(handler.Attempts, Is.EqualTo(1));
                Assert.That(File.Exists(Path.Combine(root, "policy.checkpoint")), Is.True);
            });
        }
        finally { if (Directory.Exists(root)) Directory.Delete(root, true); }
    }
}

File: Minimal/tests/Brontide.Minimal.Host.Tests/ComponentPolicyDistributionWireTests.fs
namespace Brontide.Minimal.Host.Tests

open System
open System.IO
open System.Net
open System.Net.Http
open System.Net.Http.Headers
open System.Security.Cryptography
open System.Text.Json
open System.Threading
open System.Threading.Tasks
open Brontide.Minimal.Host
open NUnit.Framework

type private Cbi40Handler(send: HttpRequestMessage -> CancellationToken -> Task<HttpResponseMessage>) =
    inherit HttpMessageHandler()
    let mutable attempts = 0
    member _.Attempts = attempts
    override _.SendAsync(request, cancellationToken) =
        attempts <- attempts + 1
        send request cancellationToken

type private Cbi40UnknownLengthContent(bytes: byte array) =
    inherit HttpContent()
    override _.SerializeToStreamAsync(stream, _) = stream.WriteAsync(bytes).AsTask()
    override _.TryComputeLength(length: byref<int64>) = length <- 0L; false
    override _.CreateContentReadStreamAsync() = Task.FromResult<Stream>(new MemoryStream(bytes, false))

type private Cbi40Observation = { Code: string; Update: bool; Attempts: int; Sha256: string option }

[<TestFixture>]
type ComponentPolicyDistributionWireTests() =
    let required (value: string | null) =
        match value with null -> failwith "A CBI40 fixture value was missing." | present -> present

    let fixture () = JsonDocument.Parse(File.ReadAllText(Path.Combine(
        TestContext.CurrentContext.TestDirectory, "component-management", "fixtures",
        "cbi40-policy-distribution-wire-vectors.json")))

    let request () : ProviderPublisherTrustPolicyDistributionRequest =
        { Challenge = String('A', 64); CurrentSequence = 0L; CurrentPolicyIdentity = None }

    let policy () =
        let entries = [ { PublisherKeyId = ProviderPublisherKeyId.create (String('A', 64)); Disposition = Admitted } ]
        { Identity = ProviderPublisherTrustPolicyIdentity.compute entries; Entries = entries }

    let signPolicy (authority: ECDsa) =
        let selected = policy ()
        let publicKey = authority.ExportSubjectPublicKeyInfo()
        let signature = authority.SignData(
            ProviderPublisherTrustPolicyUpdateManifest.encode 1L None selected.Identity,
            HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence)
        { Sequence = 1L; PreviousPolicyIdentity = None; Policy = selected; Algorithm = "ECDSA-P256-SHA256"
          AuthorityPublicKeySpkiBase64 = Convert.ToBase64String publicKey
          SignatureBase64 = Convert.ToBase64String signature }

    let wireUpdate () =
        let selected = policy ()
        { Sequence = 1L; PreviousPolicyIdentity = None; Policy = selected; Algorithm = "ECDSA-P256-SHA256"
          AuthorityPublicKeySpkiBase64 = "authority-key"; SignatureBase64 = "authority-signature" }

    let response update : ProviderPublisherTrustPolicyDistributionResponse =
        { Challenge = String('A', 64); CurrentSequence = 0L; CurrentPolicyIdentity = None
          IssuedAtUnixSeconds = 1800000000L; ExpiresAtUnixSeconds = 1800000060L; Update = update
          Algorithm = "ECDSA-P256-SHA256"; EndpointPublicKeySpkiBase64 = "endpoint-key"
          SignatureBase64 = "endpoint-signature" }

    let httpResponse (status: HttpStatusCode) (endpoint: Uri) (content: HttpContent) =
        new HttpResponseMessage(
            status,
            RequestMessage = new HttpRequestMessage(HttpMethod.Post, endpoint),
            Content = content)

    let contentOf (message: HttpRequestMessage) = message.Content |> Option.ofObj |> Option.get

    let run (vector: JsonElement) = task {
        let mutation = vector.GetProperty("mutation").GetString() |> required
        let update = wireUpdate ()
        if mutation = "request" then
            let bytes = request () |> ProviderPublisherTrustPolicyDistributionWireCodec.encodeRequest
            let decoded = bytes |> ProviderPublisherTrustPolicyDistributionWireCodec.decodeRequest
            return { Code = if decoded = request () then "request-roundtrip" else "wire-invalid"
                     Update = false; Attempts = 0; Sha256 = Some(bytes |> SHA256.HashData |> Convert.ToHexString) }
        elif mutation = "current" || mutation = "update" then
            let bytes = response (if mutation = "update" then Some update else None)
                        |> ProviderPublisherTrustPolicyDistributionWireCodec.encodeResponse
            let decoded = bytes |> ProviderPublisherTrustPolicyDistributionWireCodec.decodeResponse
            return { Code = if decoded.Update.IsSome then "response-update" else "response-current"
                     Update = decoded.Update.IsSome; Attempts = 0
                     Sha256 = Some(bytes |> SHA256.HashData |> Convert.ToHexString) }
        elif [ "truncated"; "trailing"; "marker"; "utf8" ] |> List.contains mutation then
            let original = response None |> ProviderPublisherTrustPolicyDistributionWireCodec.encodeResponse
            let bytes =
                match mutation with
                | "truncated" -> original[.. original.Length / 2 - 1]
                | "trailing" -> Array.append original [| 0uy |]
                | _ -> Array.copy original
            if mutation = "marker" then bytes[4] <- bytes[4] ^^^ 1uy
            if mutation = "utf8" then bytes[4] <- 0xFFuy
            let code =
                try ProviderPublisherTrustPolicyDistributionWireCodec.decodeResponse bytes |> ignore; "wire-accepted"
                with :? InvalidDataException -> "wire-invalid"
            return { Code = code; Update = false; Attempts = 0; Sha256 = None }
        else
            let endpoint = Uri "https://policy.example.test/v1/trust"
            let encoded = response None |> ProviderPublisherTrustPolicyDistributionWireCodec.encodeResponse
            let handler = new Cbi40Handler(fun message cancellationToken -> task {
                if mutation = "canceled" then do! Task.Delay(TimeSpan.FromSeconds 1.0, cancellationToken)
                let requestContent = contentOf message
                let! requestBytes = requestContent.ReadAsByteArrayAsync cancellationToken
                Assert.Multiple(Action(fun () ->
                    Assert.That(message.Method, Is.EqualTo(HttpMethod.Post))
                    Assert.That(message.RequestUri, Is.EqualTo(endpoint))
                    Assert.That(requestContent.Headers.ContentType |> Option.ofObj |> Option.map _.MediaType,
                        Is.EqualTo(Some HttpProviderPublisherTrustPolicyDistributionSource.MediaType))
                    Assert.That(message.Headers.Accept |> Seq.exactlyOne |> _.MediaType,
                        Is.EqualTo(HttpProviderPublisherTrustPolicyDistributionSource.MediaType))
                    Assert.That(ProviderPublisherTrustPolicyDistributionWireCodec.decodeRequest requestBytes,
                        Is.EqualTo(request ()))))
                let content: HttpContent =
                    match mutation with
                    | "declared-oversize" ->
                        new ByteArrayContent(Array.zeroCreate (ProviderPublisherTrustPolicyDistributionWireCodec.MaximumMessageBytes + 1))
                    | "streamed-oversize" ->
                        new Cbi40UnknownLengthContent(
                            Array.zeroCreate (ProviderPublisherTrustPolicyDistributionWireCodec.MaximumMessageBytes + 1))
                    | _ -> new ByteArrayContent(encoded)
                content.Headers.ContentType <- MediaTypeHeaderValue(
                    if mutation = "content-type" then "application/octet-stream"
                    else HttpProviderPublisherTrustPolicyDistributionSource.MediaType)
                if mutation = "content-encoding" then content.Headers.ContentEncoding.Add "gzip"
                let responseEndpoint = if mutation = "redirect" then Uri "https://other.example.test/v1/trust" else endpoint
                return httpResponse
                    (if mutation = "status" then HttpStatusCode.ServiceUnavailable else HttpStatusCode.OK)
                    responseEndpoint content })
            use client = new HttpClient(handler)
            let source = HttpProviderPublisherTrustPolicyDistributionSource(client, endpoint)
            use cancellation = new CancellationTokenSource()
            if mutation = "canceled" then cancellation.Cancel()
            try
                let! actual = (source :> IProviderPublisherTrustPolicyDistributionSource)
                                  .FetchAsync(request (), cancellation.Token)
                return { Code = "transport-success"; Update = actual.Update.IsSome; Attempts = handler.Attempts; Sha256 = None }
            with
            | :? OperationCanceledException ->
                return { Code = "transport-canceled"; Update = false; Attempts = handler.Attempts; Sha256 = None }
            | :? InvalidDataException ->
                return { Code = "transport-invalid"; Update = false; Attempts = handler.Attempts; Sha256 = None }
    }

    [<Test>]
    member _.``shared CBI40 vectors encode and transport only strict bounded messages``() = task {
        use document = fixture ()
        for vector in document.RootElement.GetProperty("vectors").EnumerateArray() do
            let! actual = run vector
            Assert.Multiple(Action(fun () ->
                Assert.That(actual.Code, Is.EqualTo(vector.GetProperty("code").GetString()))
                Assert.That(actual.Update, Is.EqualTo(vector.GetProperty("update").GetBoolean()))
                Assert.That(actual.Attempts, Is.LessThanOrEqualTo(1))
                match vector.TryGetProperty "sha256" with
                | true, digest -> Assert.That(actual.Sha256, Is.EqualTo(Some(digest.GetString())))
                | _ -> ()))
    }

    [<Test>]
    member _.``CBI40 C1 request codec is canonical complete and strict``() =
        let bytes = request () |> ProviderPublisherTrustPolicyDistributionWireCodec.encodeRequest
        Assert.That(bytes |> ProviderPublisherTrustPolicyDistributionWireCodec.decodeRequest
                          |> ProviderPublisherTrustPolicyDistributionWireCodec.encodeRequest,
                    Is.EqualTo(bytes :> obj))
        Assert.Throws<InvalidDataException>(Action(fun () ->
            Array.append bytes [| 0uy |] |> ProviderPublisherTrustPolicyDistributionWireCodec.decodeRequest |> ignore))
        |> ignore

    [<Test>]
    member _.``CBI40 C2 response codec preserves the complete optional update``() =
        use authority = ECDsa.Create(ECCurve.NamedCurves.nistP256)
        let update = signPolicy authority
        let decoded = response (Some update) |> ProviderPublisherTrustPolicyDistributionWireCodec.encodeResponse
                      |> ProviderPublisherTrustPolicyDistributionWireCodec.decodeResponse
        Assert.That(ProviderPublisherTrustPolicyDistributionManifest.updateDigest decoded.Update.Value,
                    Is.EqualTo(ProviderPublisherTrustPolicyDistributionManifest.updateDigest update))

    [<Test>]
    member _.``CBI40 C3 malformed truncated and trailing messages are refused``() = task {
        use document = fixture ()
        for index in [ 3; 4; 5; 6 ] do
            let! actual = run (document.RootElement.GetProperty("vectors")[index])
            Assert.That(actual.Code, Is.EqualTo("wire-invalid"))
    }

    [<Test>]
    member _.``CBI40 C4 transport requires an exact HTTPS endpoint``() =
        use client = new HttpClient(new Cbi40Handler(fun _ _ -> raise (AssertionException "not called")))
        Assert.Throws<ArgumentException>(Action(fun () ->
            HttpProviderPublisherTrustPolicyDistributionSource(client, Uri "http://policy.example.test") |> ignore))
        |> ignore

    [<Test>]
    member _.``CBI40 C5 HTTP metadata and streaming are bounded without retry``() = task {
        use document = fixture ()
        let vectors = document.RootElement.GetProperty "vectors"
        for index in [ 7; 8; 9; 10; 11; 13; 14 ] do
            let! actual = run vectors[index]
            Assert.That(actual.Attempts, Is.EqualTo(1))
    }

    [<Test>]
    member _.``CBI40 C6 HTTPS source composes with authenticated durable distribution``() = task {
        let root = Path.Combine(Path.GetTempPath(), $"brontide-cbi40-{Guid.NewGuid():N}")
        try
            use authority = ECDsa.Create(ECCurve.NamedCurves.nistP256)
            use endpointKey = ECDsa.Create(ECCurve.NamedCurves.nistP256)
            let authorityIdentity =
                authority.ExportSubjectPublicKeyInfo() |> SHA256.HashData |> Convert.ToHexString
                |> ProviderPublisherTrustPolicyAuthorityId.create
            let endpointIdentity =
                endpointKey.ExportSubjectPublicKeyInfo() |> SHA256.HashData |> Convert.ToHexString
                |> ProviderPublisherTrustPolicyDistributionEndpointId.create
            let _, opened, _ = DurableProviderPublisherTrustPolicyRegistry.Open(
                Path.Combine(root, "policy.checkpoint"), authorityIdentity, None)
            let durable = opened.Value
            let endpoint = Uri "https://policy.example.test/v1/trust"
            let now = DateTimeOffset.FromUnixTimeSeconds 1800000000L
            let update = signPolicy authority
            let handler = new Cbi40Handler(fun message cancellationToken -> task {
                let! bytes = (contentOf message).ReadAsByteArrayAsync cancellationToken
                let decoded = ProviderPublisherTrustPolicyDistributionWireCodec.decodeRequest bytes
                let publicKey = endpointKey.ExportSubjectPublicKeyInfo()
                let issued, expires = now.ToUnixTimeSeconds(), now.AddMinutes(1.0).ToUnixTimeSeconds()
                let signature = endpointKey.SignData(
                    ProviderPublisherTrustPolicyDistributionManifest.encode decoded.Challenge decoded.CurrentSequence
                        decoded.CurrentPolicyIdentity issued expires (Some update),
                    HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence)
                let envelope =
                    { Challenge = decoded.Challenge; CurrentSequence = decoded.CurrentSequence
                      CurrentPolicyIdentity = decoded.CurrentPolicyIdentity; IssuedAtUnixSeconds = issued
                      ExpiresAtUnixSeconds = expires; Update = Some update; Algorithm = "ECDSA-P256-SHA256"
                      EndpointPublicKeySpkiBase64 = Convert.ToBase64String publicKey
                      SignatureBase64 = Convert.ToBase64String signature }
                let content = new ByteArrayContent(
                    ProviderPublisherTrustPolicyDistributionWireCodec.encodeResponse envelope)
                content.Headers.ContentType <- MediaTypeHeaderValue(
                    HttpProviderPublisherTrustPolicyDistributionSource.MediaType)
                return httpResponse HttpStatusCode.OK endpoint content })
            use httpClient = new HttpClient(handler)
            let source = HttpProviderPublisherTrustPolicyDistributionSource(httpClient, endpoint)
            let client = ProviderPublisherTrustPolicyDistributionClient(durable, endpointIdentity)
            let! result = client.SynchronizeAsync(source, now, TimeSpan.FromSeconds 1.0, CancellationToken.None)
            Assert.Multiple(Action(fun () ->
                Assert.That(result.IsApplied, Is.True)
                Assert.That(result.Floor.Sequence, Is.EqualTo(1L))
                Assert.That(handler.Attempts, Is.EqualTo(1))
                Assert.That(File.Exists(Path.Combine(root, "policy.checkpoint")), Is.True)))
        finally
            if Directory.Exists root then Directory.Delete(root, true)
    }
